File: Submission 3/ghidra_script/ExtractOpcode.py
# ExtractOpcodes.py
# Extracts opcode mnemonics from all instructions in the current program
# and saves them in multiple formats for ML / analysis use.

from ghidra.program.model.listing import Instruction
from ghidra.util.task import TaskMonitor
from java.io import FileWriter, BufferedWriter
from java.util import ArrayList
import json
import os

program = currentProgram
listing = program.getListing()

# Ask user for output directory
output_dir = askDirectory("Select output directory", "Choose")

# Ask for base filename
base_name = askString("Sample name", "Enter base filename (no .exe)")

opcode_file = os.path.join(output_dir.getAbsolutePath(), base_name + ".opcode")
opcode_hex_file = os.path.join(output_dir.getAbsolutePath(), base_name + ".opcode_hex")
meta_file = os.path.join(output_dir.getAbsolutePath(), base_name + ".meta.json")

mnemonics = []
mnemonics_hex = []

instr_iter = listing.getInstructions(True)

count = 0
while instr_iter.hasNext():
    instr = instr_iter.next()
    mnem = instr.getMnemonicString()
    mnemonics.append(mnem)

    bytes_hex = instr.getBytes()
    hex_str = ''.join(['%02x' % (b & 0xff) for b in bytes_hex])
    mnemonics_hex.append(hex_str)

    count += 1

# Write opcode mnemonics
with open(opcode_file, "w") as f:
    for m in mnemonics:
        f.write(m + "\n")

# Write opcode hex
with open(opcode_hex_file, "w") as f:
    for h in mnemonics_hex:
        f.write(h + "\n")

# Metadata
meta = {
    "sample": base_name,
    "instruction_count": count,
    "language": str(program.getLanguage()),
    "compiler": str(program.getCompilerSpec()),
    "ghidra_version": getGhidraVersion()
}

with open(meta_file, "w") as f:
    json.dump(meta, f, indent=2)

print("[+] Opcode extraction complete")
print("[+] Instructions:", count)

File: Submission 3/ghidra_script/ExtractOpcode.java
// ExtractOpcode.java
// Exports instruction mnemonics for the entire program to <outDir>\<programName>.opcode
// Headless usage: -postScript ExtractOpcode "<outDir>"

import java.io.*;
import java.nio.charset.StandardCharsets;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.Listing;

public class ExtractOpcode extends GhidraScript {

    @Override
    public void run() throws Exception {
        if (currentProgram == null) {
            println("No currentProgram. Exiting.");
            return;
        }

        String outDir = null;
        if (getScriptArgs() != null && getScriptArgs().length > 0) {
            outDir = getScriptArgs()[0];
        }

        if (outDir == null || outDir.trim().isEmpty()) {
            printerr("ERROR: Output directory not provided. Usage: -postScript ExtractOpcode \"C:\\path\\OpCode\"");
            return;
        }

        File outFolder = new File(outDir);
        if (!outFolder.exists() && !outFolder.mkdirs()) {
            printerr("ERROR: Could not create output directory: " + outFolder.getAbsolutePath());
            return;
        }

        // Program name without extension (so your output matches your EXE base name)
        String baseName = currentProgram.getName();
        int dot = baseName.lastIndexOf('.');
        if (dot > 0) baseName = baseName.substring(0, dot);

        File outFile = new File(outFolder, baseName + ".opcode");

        // Skip overwriting a non-empty existing opcode file
        if (outFile.exists() && outFile.length() > 0) {
            println("Skipping (already exists, non-empty): " + outFile.getAbsolutePath());
            return;
        }

        Listing listing = currentProgram.getListing();
        long count = 0;

        // Iterate ALL instructions in the program (not functions, not decompiler)
        try (BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
                new FileOutputStream(outFile, false), StandardCharsets.UTF_8))) {

            for (Instruction ins : listing.getInstructions(true)) {
                Address a = ins.getAddress();
                String mnem = ins.getMnemonicString();
                if (mnem != null && !mnem.isEmpty()) {
                    // If you want just opcode mnemonic:
                    bw.write(mnem);

                    // If you want address + mnemonic instead, use:
                    // bw.write(a.toString() + " " + mnem);

                    bw.newLine();
                    count++;
                }

                if (monitor.isCancelled()) {
                    println("Cancelled by user.");
                    break;
                }
            }

            // Never leave it empty — helps you detect “no instructions were created”
            if (count == 0) {
                bw.write("NO_INSTRUCTIONS_FOUND");
                bw.newLine();
            }
        }

        println("Wrote " + count + " instructions to: " + outFile.getAbsolutePath());
    }
}